### PR TITLE
feat: cache settings objects when during upsert operations

### PR DESCRIPTION
**Pull request by warber**
#### What this PR does / Why we need it:

The current process of upserting the Settings 2.0 object is highly inefficient since it requires retrieving all settings objects to find those with a matching legacy external ID. This pull request proposes the implementation of a basic cache to eliminate the need for this extra step for every settings object being upserted.

#### Does this PR introduce a user-facing change?

deploying settings 2.0 objects is faster :rocket:

### CODE REVIEW

File: pkg/client/dtclient/client.go [hunk 1]
```diff
@@ -432,25 +435,21 @@ func (d *DynatraceClient) upsertSettings(ctx context.Context, obj SettingsObject
 	if err != nil {
 		return DynatraceEntity{}, fmt.Errorf("unable to generate external id: %w", err)
 	}
-	settingsWithExternalID, err := d.listSettings(ctx, obj.SchemaId, ListSettingsOptions{
-		Filter: func(object DownloadSettingsObject) bool {
-			return object.ExternalId == legacyExternalID
-		},
-	})
 
+	settingsWithExternalID, err := d.getSettingsFromCache(ctx, obj.SchemaId, func(object DownloadSettingsObject) bool { return object.ExternalId == legacyExternalID })
```
Comment (UnseenWizzard) [inline pkg/client/dtclient/client.go:439]: My initial thought would have been to make caching a built-in feature of ListSettings and transparent to the callsites. If I just use the public ListSettings now or listSettings in another internal method I'll have to know about and use the FromCache method explicitly. If I use the public interface I have no way to profit from caching.
What made you go for explicitly using the cache?

File: pkg/client/dtclient/settings_cache.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dtclient
+
+type SettingsCache struct {
```
Comment (UnseenWizzard) [inline pkg/client/dtclient/settings_cache.go:19]: could/should we make this a generic key-value cache? We'll need a schema and config API object cache soon as well

**Comment by warber** [comment]
@UnseenWizzard the PR is in draft an not yet ready to review :)

File: pkg/client/dtclient/client.go (hunk 1 above)
Comment (warber) [reply pkg/client/dtclient/client.go:439]: @UnseenWizzard I see the cache as something that will eventually removed again once we do not need to list all settings in upsert anymore. I don't see any other strict/important/useful reason why we need a cache everywhere, but maybe you can help me there :)

File: pkg/client/dtclient/settings_cache.go (hunk 2 above)
Comment (warber) [reply pkg/client/dtclient/settings_cache.go:19]: @UnseenWizzard  it's coming :) For now just imagine to read `T` insead of `DownloadSettingsObject` :D

File: pkg/client/dtclient/client.go (hunk 1 above)
Comment (UnseenWizzard) [reply pkg/client/dtclient/client.go]: > once we do not need to list all settings in upsert anymore.
I think with the change to look for unique key properties, this will never go away anymore.

**Comment by UnseenWizzard** [comment]
> @UnseenWizzard the PR is in draft an not yet ready to review :)

@warber if you open a draft of something I find interesting I will take a look at it :P 
Of course, feel free to ignore comments at the draft stage, especially for things that'll change anyway (but I might ask them again once it's ready).

**Comment by warber** [comment]
@UnseenWizzard ready for review.

File: pkg/client/dtclient/client.go (hunk 1 above)
Comment (warber) [reply pkg/client/dtclient/client.go]: I've changed it to be used "built-in" in `listSettings()` :)

File: pkg/client/dtclient/cache.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+package dtclient
+
+import "sync"
+
+type cache[T any] struct {
+	cachedItems map[string][]T
+	mutex       sync.RWMutex
+}
+
+func (s *cache[T]) hasCache(id string) bool {
```
Comment (Laubi) [inline pkg/client/dtclient/cache.go:26]: Minor: `isCached` or just `has`? Or even `hasCached` instead of `hasCache`

File: pkg/client/dtclient/cache.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+func (s *cache[T]) filter(id string, filter func(T) bool) []T {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	if filter == nil {
+		filter = func(object T) bool { return true }
+	}
+	result := make([]T, 0)
+	for _, i := range s.cachedItems[id] {
```
Comment (Laubi) [inline pkg/client/dtclient/cache.go:52]: This will fail if cachedItems has not been initialized yet

File: pkg/client/dtclient/cache.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dtclient
+
+import "sync"
+
+type cache[T any] struct {
+	cachedItems map[string][]T
```
Comment (Laubi) [inline pkg/client/dtclient/cache.go:22]: map-key is what? Schema ID? Please specify or add brief docs.

File: pkg/client/dtclient/cache.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	}
+	s.cachedItems[id] = settings
+}
+
+func (s *cache[T]) filter(id string, filter func(T) bool) []T {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	if filter == nil {
+		filter = func(object T) bool { return true }
```
Comment (Laubi) [inline pkg/client/dtclient/cache.go:49]: I'd argue we should just return the entries here instead of looping them later and adding them to a map.

File: pkg/client/dtclient/cache.go (hunk 5 above)
Comment (warber) [reply pkg/client/dtclient/cache.go:22]: it's **a** key.
might be schema id, depending on the use case...

File: pkg/client/dtclient/cache.go (hunk 6 above)
Comment (warber) [reply pkg/client/dtclient/cache.go:49]: haha yeah, of course :smile:

File: pkg/client/dtclient/cache.go (hunk 4 above)
Comment (warber) [reply pkg/client/dtclient/cache.go:52]: it's a good point :) , but you are wrong - ranging over or reading from a nil map will not panic.

Comment (Laubi) [reply pkg/client/dtclient/cache.go:52]: True - my error. Even accessing a  key that does not exist returns nil, thus ranging does work.

**Review by Laubi — APPROVED** [review]

File: pkg/client/dtclient/cache.go (hunk 4 above)
Comment (warber) [reply pkg/client/dtclient/cache.go:52]: In this case, yes. Generally, it returns the zero value of the map's value type. :nerd_face:

File: pkg/client/dtclient/cache.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+package dtclient
+
+import "sync"
+
+type cache[T any] struct {
+	cachedItems map[string][]T
+	mutex       sync.RWMutex
+}
+
+func (s *cache[T]) isCached(id string) bool {
```
Comment (UnseenWizzard) [inline pkg/client/dtclient/cache.go:26]: might be a stupid question / the cache usage is super specialized on filter(, but: 

Why is there no`get(id string) (T, bool)` at all? 
That would allow me to just do the _, ok check on my own where needed - following the idioms of map

File: pkg/client/dtclient/client.go [hunk 8]
```diff
@@ -432,25 +435,22 @@ func (d *DynatraceClient) upsertSettings(ctx context.Context, obj SettingsObject
...
 		return DynatraceEntity{}, fmt.Errorf("unable to generate external id: %w", err)
 	}
+
 	settingsWithExternalID, err := d.listSettings(ctx, obj.SchemaId, ListSettingsOptions{
 		Filter: func(object DownloadSettingsObject) bool {
 			return object.ExternalId == legacyExternalID
 		},
 	})
 
-	if err != nil {
```
Comment (UnseenWizzard) [inline pkg/client/dtclient/client.go:441]: why did you remove this error handling? Won't we continue now in case the GET /settings API call failed?
In the case of client error here, it's likely further calls will just error as well, but still at this point data is invalid, isn't it?

**Review by UnseenWizzard — DISMISSED** [review]
Thanks for the fix. Good to merge as far as I'm concerned, if you decide against the `get` method nitpick :)

File: pkg/client/dtclient/cache.go (hunk 7 above)
Comment (warber) [reply pkg/client/dtclient/cache.go:26]: i like the idea. will do

**Comment by warber** [comment]
@UnseenWizzard ready for review (again) :)

**Review by jskelin — CHANGES_REQUESTED** [review]

**Review by jskelin — CHANGES_REQUESTED** [review]
Using generics for values, this pretends to be a general solution for caching. On the other hand, the `get` method with integrated filtration is highly profiled for one purpose. 
Either make this solution a general solution (separate package with well-defined API). Or keep it local, non-exported and simple as possible that implement only specific demands (without generic)

**Comment by warber** [comment]
@jskelin 

> the get method with integrated filtration is highly profiled for one purpose.

not really, it's a generic `get` method with a generic `filter` that can be applied (or not, you can also just pass `nil`)

> Either make this solution a general solution (separate package with well-defined API). Or keep it local, non-exported and simple as possible that implement only specific demands (without generic)!

For now, we don't need a cache anywhere but in this package and there we need it for multiple data types. To me that looks like a perfect fit for keeping it local and generic.

File: pkg/client/dtclient/cache.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	for _, i := range s.cachedItems[id] {
+		if filter(i) {
+			result = append(result, i)
+		}
+	}
+	return result, true
+
+}
+
+func (s *cache[T]) invalidate(id string) {
```
[the remark this answers is not in the document]
Comment (warber) [reply pkg/client/dtclient/cache.go:66]: could be done. however, i'd argure that it's also prette clear what the invalidate method might do in the context of a cache.
I am also fine with `delete' tho.

File: pkg/client/dtclient/cache.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+func (s *cache[T]) isCached(id string) bool {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+
+	_, ok := s.cachedItems[id]
+	return ok
+}
+
+func (s *cache[T]) set(id string, settings []T) {
```
[the remark this answers is not in the document]
Comment (warber) [reply pkg/client/dtclient/cache.go:34]: agree

**Comment by warber** [comment]
@jskelin 
I've refactored a bit, I think you will love it :)

**Review by jskelin — APPROVED** [review]
looks great

**Review by UnseenWizzard — APPROVED** [review]
I like the new version! 🚀
